### PR TITLE
Taskgroup decorator

### LINKED ISSUE #11870: Add @taskgroup decorator
**Description**

Similarly to `@dag` and `@task` decorators, we can add a decorator to easily generate `TaskGroup` instances. 

**Use case / motivation**

This would be used for reusable pieces of DAGs. An example may be to hyperparameter tune an ML pipeline. You may want to generate several parallel subpipelines with different settings/parameters to be executed in parallel. At the end pull the result and decide the best parametrization.  

```py
@task
collect_dataset(...)
  pass

@task
def train_model(...)
   pass

@task_group
def movielens_model_pipeline(learning_rate: int, feature_list: list, dataset: XComArg):  
  dataset = filter_dataset(dataset, feature_list)
  train_model(dataset, learning_rate)

@dag
def movielens_hpt(dataset_path: str, feature_list:list=['year', 'director']):
  dataset = load_dataset(dataset_path)
  for i in range(0.1, 0.9, step=0.1):
    movielens_model_pipeline(i, feature_list, dataset)
  
  decide_best_model(...)
```

**Related Issues**

https://github.com/apache/airflow/issues?q=is%3Aopen+is%3Aissue+label%3AAIP-31

### PR DESCRIPTION
**Pull request by dimberman**
From https://github.com/apache/airflow/pull/13479

Added task_group decorator which can be used to create taskgroup from python callable. Inside python callable tasks can be grouped by calling tasks (python callable ) created with task decorator.

- taskgroup decorator takes both optional and keyword argument which are passed to Constructor of TaskGroup class.

- TaskGroup can be created with one of following syntax

``` python
@task_group
@task_group()
@task_group(group_id='group_name')
```

- TaskGroup class  constructor takes one mandatory argument group_id, if not given in decorator it sets group_id to python callable name.

Following is a simple example demonstrating use of taskgroup decorator grouping multiple tasks.

``` python
@task
def task_1(value):
  return f'[ Task1 {value} ]'


@task
def task_2(value):
  print(f'[ Task2 {value} ]')


@task_group
def section_1(value):
  return task_2(task_1(value))
```

task_group decorator utilizes existing TaskGroup context manager currently used for creating TaskGroup, which means we can create nested taskgroup by created nested callable. Following is an example demonstrating use of nested taskgroup.

``` python
@task
def task_start():
    return '[Task_start]'

@task
def task_end():
    print(f'[ Task_End ]')

@task
def task_1(value):
    return f'[ Task1 {value} ]'

@task
def task_2(value):
    print(f'[ Task2 {value} ]')

@task
def task_3(value):
    return f'[ Task3 {value} ]'

@task
def task_4(value):
    print(f'[ Task4 {value} ]')

@task_group
def section_1(value):

    @task_group
    def section_2(value2):
        return task_4(task_3(value2))

    op1 = task_2(task_1(value))
    return section_2(op1) 
```

Dedicated test cases for taskgroup decorator is created in file 
/tests/utils/test_task_group_decorator.py

Recent changes 

- Added logic to append suffix to duplicate group_id.

closes: #11870

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/example_dags/example_task_group_decorator.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    """ TaskGroup for grouping related Tasks"""
+    return task_3(task_2(task_1(value)))
+
+
+# Executing Tasks and TaskGroups
+with DAG(dag_id="example_task_group_decorator", start_date=days_ago(2), tags=["example"]) as dag:
+    t1 = task_start()
+
+    s1 = section_1(t1)
+    s1.set_downstream(task_end())
```
Comment (ashb) [inline airflow/example_dags/example_task_group_decorator.py]: Can you do this as `s1 >> task_end()`?

File: airflow/example_dags/example_task_group_decorator.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Example DAG demonstrating the usage of the @taskgroup decorator."""
+
+from airflow.decorators import task, taskgroup
```
Comment (ashb) [inline airflow/example_dags/example_task_group_decorator.py]:
```suggestion
from airflow.decorators import task, task_group
```

etc now, I think?

File: docs/apache-airflow/concepts.rst [hunk 3]
```diff
@@ -1128,6 +1128,14 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
 .. image:: img/task_group.gif
 
 
+TaskGroup can be created using ``@taskgroup`` decorator, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class, if not given it copies function name as ``group_id``. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```
Comment (ashb) [inline docs/apache-airflow/concepts.rst]:
```suggestion
TaskGroup can be created using ``@task?group`` decorator, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class, if not given it copies function name as ``group_id``. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```

File: docs/spelling_wordlist.txt [hunk 4]
```diff
@@ -1316,6 +1316,7 @@ tagKey
 tagValue
 tao
 taskflow
+taskgroup
```
Comment (ashb) [inline docs/spelling_wordlist.txt]:
```suggestion
task_group
```

(If it's even needed cos of the underscore now?)

File: airflow/example_dags/example_task_group_decorator.py (hunk 1 above)
Comment (dimberman) [reply airflow/example_dags/example_task_group_decorator.py]: @ashb I added a better example to show how this can be used as a generator.

File: airflow/decorators/task_group.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    def wrapper(f: T):
+
+        # Setting group_id as function name if not given in kwarg group_id
+        if len(tg_args) == 0 and 'group_id' not in tg_kwargs.keys():
+            tg_kwargs['group_id'] = f.__name__
+
+        # Get dag initializer signature and bind it to validate that task_group_args,
+        # and task_group_kwargs are correct
+        task_group_sig = signature(TaskGroup.__init__)
+        task_group_bound_args = task_group_sig.bind_partial(*tg_args, **tg_kwargs)
```
Comment (ashb) [inline airflow/decorators/task_group.py]: This could all be moved outside wrapper - it doesn't change for each invocation of the decorated func.

File: airflow/decorators/task_group.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        @functools.wraps(f)
+        def factory(*args, **kwargs):
+            # Generate signature for decorated function and bind the arguments when called
+            # we do this to extract parameters so we can annotate them on the DAG object.
+            # In addition, this fails if we are missing any args/kwargs with TypeError as expected.
+            f_sig = signature(f).bind(*args, **kwargs)
+            # Apply defaults to capture default values if set.
+            f_sig.apply_defaults()
```
Comment (ashb) [inline airflow/decorators/task_group.py]: And could this be moved outside factory?

Comment (dimberman) [reply airflow/decorators/task_group.py]: @ashb this can't be moved outside of the factory because it needs the *args and **kwargs

File: airflow/decorators/task_group.py (hunk 5 above)
Comment (dimberman) [reply airflow/decorators/task_group.py]: @ashb the first section relies on f.__name__ so that does change. The second section can be moved.

File: airflow/decorators/task_group.py (hunk 6 above)
Comment (ashb) [reply airflow/decorators/task_group.py]: Perhaps just move the `f_sig = signature(f)` to outside of the factory - when it's called once it will make no difference, but if the TG is used more than once getting the signatures are relatively costly.

File: airflow/utils/task_group.py [hunk 7]
```diff
@@ -95,8 +95,22 @@ def __init__(
...
+                    int(re.split(r'^.+__', used_group_id)[1])
+                    for used_group_id in self.used_group_ids
+                    if used_group_id is not None and re.match(rf'^{base}__\d+$', used_group_id)
+                ]
+            )
+            if not suffixes:
+                self._group_id += '__1'
+            else:
+                self._group_id = f'{base}__{suffixes[-1] + 1}'
+
```
Comment (ashb) [inline airflow/utils/task_group.py]: Should this only apply to the decorator? IF you aren't using the decorator function and you try to reuse a TG id that kind of feels like it should be an error.

(I don't have a strong opinion either way.)

Comment (dimberman) [reply airflow/utils/task_group.py]: @ashb I'm going to add a "from_decorator" parameter. I agree that if not in a decorator this should fail.

File: airflow/decorators/task_group.py (hunk 6 above)
Comment (dimberman) [reply airflow/decorators/task_group.py]: I moved it as high up as I could.

File: airflow/decorators/task_group.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+"""
+import functools
+from inspect import signature
+from typing import TYPE_CHECKING, Callable, Optional, TypeVar, cast
+
+from airflow.utils.task_group import TaskGroup
+
+if TYPE_CHECKING:
+    pass
+
```
Comment (ashb) [inline airflow/decorators/task_group.py]:
```suggestion
```

File: airflow/utils/task_group.py [hunk 9]
```diff
@@ -54,6 +54,8 @@ class TaskGroup(TaskMixin):
     :type ui_color: str
     :param ui_fgcolor: The label color of the TaskGroup node when displayed in the UI
     :type ui_fgcolor: str
+    :param from_decorator: was this taskgroup created by a taskgroup decorator (should only be set by Airflow)
+    :type from_decorator: bool
```
Comment (ashb) [inline airflow/utils/task_group.py]:
```suggestion
    :param _from_decorator: was this taskgroup created by a taskgroup decorator (should only be set by Airflow)
    :type _from_decorator: bool
```


Or perhaps better might be:

```suggestion
    :param add_suffix_on_collision: If this task group name already exists, automatically add `__1` etc suffixes
    :type from_decorator: add_suffix_on_collision
```

The idea here is that "from_decorator" says nothing about what this parameter does, but add_suffix_on_collision should be immediately clear, especially at the caller.

File: airflow/utils/task_group.py [hunk 10]
```diff
@@ -304,7 +323,7 @@ def build_map(task_group):
         build_map(self)
         return task_group_map
 
-    def get_child_by_label(self, label: str) -> Union["BaseOperator", "TaskGroup"]:
+    def get_child_by_label(self, label: str) -> Union["BaseOperator", "TaskGroup" ""]:
```
Comment (ashb) [inline airflow/utils/task_group.py]:
```suggestion
    def get_child_by_label(self, label: str) -> Union["BaseOperator", "TaskGroup"]:
```

File: tests/utils/test_task_group.py [hunk 11]
```diff
@@ -525,42 +527,6 @@ def test_dag_edges():
...
-            with TaskGroup("group1"):
-                _ = DummyOperator(task_id="downstream_join_id")
-
-    with pytest.raises(DuplicateTaskIdFound, match=r".* 'group1.upstream_join_id' .*"):
-        with DAG("test_duplicate_group_id", start_date=execution_date):
-            _ = DummyOperator(task_id="task1")
-            with TaskGroup("group1"):
-                _ = DummyOperator(task_id="upstream_join_id")
-
-
```
Comment (ashb) [inline tests/utils/test_task_group.py]: I think you now want to keep this entire block

File: docs/apache-airflow/concepts.rst [hunk 12]
```diff
@@ -1128,6 +1128,14 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
 .. image:: img/task_group.gif
 
 
+TaskGroup can be created using ``@task_group`` decorator, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class, if not given it copies function name as ``group_id``. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```
Comment (ashb) [inline docs/apache-airflow/concepts.rst]: It's not mandatory if it works out a default ;)

File: airflow/decorators/task_group.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    :param tg_kwargs: Kwargs for TaskGroup object.
+    :type tg_kwargs: dict
+    """
+    # Setting group_id as function name if not given in kwarg group_id
+
+    # Get dag initializer signature and bind it to validate that task_group_args,
+    # and task_group_kwargs are correct
+
+    def wrapper(f: T):
+        if len(tg_args) == 0 and 'group_id' not in tg_kwargs.keys():
```
Comment (ashb) [inline airflow/decorators/task_group.py]:
```suggestion
        if not tg_args and 'group_id' not in tg_kwargs:
```

File: tests/utils/test_task_group.py [hunk 14]
```diff
@@ -576,3 +579,372 @@ def test_task_without_dag():
...
+    def task_group1():
+        task_start()
+        task_1()
+        task_2()
+
+    @task_group_decorator(group_id='task_group1')
+    def task_group2():
+        task_3()
+
+    @task_group_decorator(group_id='task_group1')
```
Comment (ashb) [inline tests/utils/test_task_group.py:867]:
```suggestion
    @task_group_decorator('task_group1')
```

So we test passing group id via arg (instead of kwarg)

File: tests/utils/test_task_group.py [hunk 15]
```diff
@@ -576,3 +579,372 @@ def test_task_without_dag():
...
+    def task_group3():
+        task_end()
+
+    execution_date = pendulum.parse("20201109")
+    with DAG(dag_id="example_duplicate_task_group_id", start_date=execution_date, tags=["example"]) as dag:
+        task_group1()
+        task_group2()
```
Comment (ashb) [inline tests/utils/test_task_group.py:874]: I would have expected this to fail: this is three _separate_ task groups, all with the same name.

Calling the same TG more than once is when I would expect the `__1` suffix to be added, but having _actual_ duplicate TG ids for different groups I think should be an error case.

File: airflow/decorators/task_group.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        f_sig = signature(f)
+
+        @functools.wraps(f)
+        def factory(*args, **kwargs):
+            # Generate signature for decorated function and bind the arguments when called
+            # we do this to extract parameters so we can annotate them on the DAG object.
+            # In addition, this fails if we are missing any args/kwargs with TypeError as expected.
+            # Apply defaults to capture default values if set.
+            current_f_sig = f_sig.bind(*args, **kwargs)
+            current_f_sig.apply_defaults()
```
Comment (ashb) [inline airflow/decorators/task_group.py]: I'm a bit confused as to why we need this at all, and why we can't simply do `f(*args, **kwargs)` on L69.

The comment says "so we can annotate them on the DAG object." but I don't see any reference to that in the code.

File: airflow/decorators/task_group.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    :param python_callable: Function to decorate
+    :param tg_args: Arguments for TaskGroup object
+    :type tg_args: list
+    :param tg_kwargs: Kwargs for TaskGroup object.
+    :type tg_kwargs: dict
+    """
+    # Setting group_id as function name if not given in kwarg group_id
+
+    # Get dag initializer signature and bind it to validate that task_group_args,
+    # and task_group_kwargs are correct
```
Comment (ashb) [inline airflow/decorators/task_group.py]: These comments are now out of place.

**Review by ashb — CHANGES_REQUESTED** [review]
Not quite sure the duplicate/reusing TG id behaviour is right yet, and code can be simplified a bit I think.

File: airflow/decorators/task_group.py (hunk 16 above)
Comment (dimberman) [reply airflow/decorators/task_group.py]: @ashb I don't know either, that was there when I ported it. I don't see anything like that in the other decorators and the tests still pass without it so I'm going to remove it.

File: tests/utils/test_task_group.py (hunk 15 above)
Comment (dimberman) [reply tests/utils/test_task_group.py:874]:
```
    node_ids = {
        'id': None,
        'children': [
            {
                'id': 'task_group1',
                'children': [
                    {'id': 'task_group1.start_task'},
                    {'id': 'task_group1.task'},
                    {'id': 'task_group1.task__1'},
                ],
            },
            {'id': 'task_group1__1', 'children': [{'id': 'task_group1__1.task1'}]},
            {'id': 'task_group1__2', 'children': [{'id': 'task_group1__2.end_task'}]},
        ],
    }
```

I think this is the correct functionality. With taskgroup decorators people might be importing these from external sources and might therefore not have control over the group_id of that TaskGroup. Automatically adding a suffix can prevent failed DAGs when using these externally created resources.

Comment (ashb) [reply tests/utils/test_task_group.py:874]: Hmmm, given there's no way to override the name of a TG, true.

**Review by ashb — APPROVED** [review]
